openforcefield/tests/test_molecule.py

```diff
@@ -529,6 +530,326 @@ def test_name(self):
+        # now test all of the variations
+        assert Molecule.are_isomorphic(benzene, benzene_no_aromatic, aromatic_matching=inputs['aromatic_matching'],
+                                       formal_charge_matching=inputs['formal_charge_matching'],
+                                       bond_order_matching=inputs['bond_order_matching'],
+                                       atom_stereochemistry_matching=inputs['atom_stereochemistry_matching'],
+                                       bond_stereochemistry_matching=inputs['bond_stereochemistry_matching'])[0] is inputs['result']
+
+    def test_remap(self):
+        """Test the remap function which should return a new molecule in the requested ordering"""
+        # the order here is CCO
+        ethanol = Molecule.from_file(get_data_file_path('molecules/ethanol.sdf'))
+        # get ethanol in reverse order OCC
+        ethanol_reverse = create_reversed_ethanol()
+        # get the mapping between the molecules
+        mapping = Molecule.are_isomorphic(ethanol, ethanol_reverse, True)[1]
+        ethanol.add_bond_charge_virtual_site([0, 1], 0.3 * unit.angstrom)
+        # make sure that molecules with virtual sites raises an error
+        with pytest.raises(NotImplementedError):
+            remapped = ethanol.remap(mapping, current_to_new=True)
+
+        # remake with no virtual site and remap to match the reversed ordering
+        ethanol = Molecule.from_file(get_data_file_path('molecules/ethanol.sdf'))
+
+        new_ethanol = ethanol.remap(mapping, current_to_new=True)
+
+        def assert_molecules_match_after_remap(mol1, mol2):
+            """Check all of the attributes in a molecule match after being remapped"""
+            for atoms in zip(mol1.atoms, mol2.atoms):
+                assert atoms[0].to_dict() == atoms[1].to_dict()
+            # bonds will not be in the same order in the molecule and the atom1 and atom2 indecies could be out of order
+            # make a dict to compare them both
+            remaped_bonds = dict(((bond.atom1_index, bond.atom2_index), bond) for bond in mol2.bonds)
+            for bond in mol1.bonds:
+                key = (bond.atom1_index, bond.atom2_index)
+                if key not in remaped_bonds:
+                    key = tuple(reversed(key))
+                assert key in remaped_bonds
+                # now compare each attribute of the bond except the atom indexes
+                bond_dict = bond.to_dict()
+                del bond_dict['atom1']
+                del bond_dict['atom2']
+                remapped_bond_dict = remaped_bonds[key].to_dict()
+                del remapped_bond_dict['atom1']
+                del remapped_bond_dict['atom2']
+            assert mol1.n_bonds == mol2.n_bonds
+            assert mol1.n_angles == mol2.n_angles
+            assert mol1.n_propers == mol2.n_propers
+            assert mol1.n_impropers == mol2.n_impropers
+            assert mol1.total_charge == mol2.total_charge
```

Comment:
I have added this to the check as well but changed the construction of the normal ethanol to use the toolkit independent function instead of the sdf file to make sure the charges are assigned.

Reply:
We might want to update this test when we can read charges from SDF files?